From c9405ca1e50572e41ec56561c23cc2ac516d1a47 Mon Sep 17 00:00:00 2001
From: Peter Jung <admin@ptr1337.dev>
Date: Wed, 1 Nov 2023 20:30:44 +0100
Subject: [PATCH] Add-extra-version-CachyOS

Signed-off-by: Peter Jung <admin@ptr1337.dev>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 5c418efbe89b6..dd8ce68271ab0 100644
--- a/Makefile
+++ b/Makefile
@@ -6 +6 @@
-EXTRAVERSION =
+EXTRAVERSION = -cachyos
