From 11a885bbecd7b9862de23bf414fb422fbc4ec19c Mon Sep 17 00:00:00 2001
From: Peter Jung <admin@ptr1337.dev>
Date: Thu, 29 Dec 2022 12:12:32 +0100
Subject: [PATCH 12/25] Cachy: mm: Disable unevictable compaction

Signed-off-by: Peter Jung <admin@ptr1337.dev>
---
 mm/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mm/Kconfig b/mm/Kconfig
index b1448aa81..64da441d5 100644
--- a/mm/Kconfig
+++ b/mm/Kconfig
@@ -614,7 +614,7 @@ config COMPACTION
 config COMPACT_UNEVICTABLE_DEFAULT
 	int
 	depends on COMPACTION
-	default 0 if PREEMPT_RT
+	default 0 if PREEMPT_RT || CACHY
 	default 1
 
 #
-- 
2.43.0.232.ge79552d197

