From f34f5dd19cf4d6d4686777cc631c52f6d1a4f1ce Mon Sep 17 00:00:00 2001
From: "Jan Alexander Steffens (heftig)" <heftig@archlinux.org>
Date: Mon, 12 Dec 2022 00:03:03 +0100
Subject: [PATCH 21/25] Cachy: Use Kyber as the elevator for MQ devices

---
 block/elevator.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/block/elevator.c b/block/elevator.c
index 8f8baa90b..1c93fe91b 100644
--- a/block/elevator.c
+++ b/block/elevator.c
@@ -574,7 +574,13 @@ static struct elevator_type *elevator_get_default(struct request_queue *q)
 
 	if (q->nr_hw_queues != 1 &&
 	    !blk_mq_is_shared_tags(q->tag_set->flags))
+#if defined(CONFIG_CACHY) && defined(CONFIG_MQ_IOSCHED_KYBER)
+		return elevator_find_get(q, "kyber");
+#elif defined(CONFIG_CACHY)
+		return elevator_find_get(q, "mq-deadline");
+#else
 		return NULL;
+#endif
 
 #if defined(CONFIG_CACHY) && defined(CONFIG_IOSCHED_BFQ)
 	return elevator_find_get(q, "bfq");
-- 
2.43.0.232.ge79552d197

