diff -uNr linux-5.10.57/drivers/hid/usbhid/hid-core.c linux-5.10.57-new/drivers/hid/usbhid/hid-core.c
--- linux-5.10.57/drivers/hid/usbhid/hid-core.c	2021-08-20 10:29:46.395739447 +0100
+++ linux-5.10.57-new/drivers/hid/usbhid/hid-core.c	2021-08-20 10:29:21.756271000 +0100
@@ -57,6 +57,10 @@
 module_param_named(kbpoll, hid_kbpoll_interval, uint, 0644);
 MODULE_PARM_DESC(kbpoll, "Polling interval of keyboards");
 
+static unsigned int hid_poll_interval;
+module_param_named(poll, hid_poll_interval, uint, 0644);
+MODULE_PARM_DESC(poll, "Default polling rate for all devices");
+
 static unsigned int ignoreled;
 module_param_named(ignoreled, ignoreled, uint, 0644);
 MODULE_PARM_DESC(ignoreled, "Autosuspend with active leds");
@@ -1112,21 +1116,25 @@
 		/* Change the polling interval of mice, joysticks
 		 * and keyboards.
 		 */
-		switch (hid->collection->usage) {
-		case HID_GD_MOUSE:
-			if (hid_mousepoll_interval > 0)
-				interval = hid_mousepoll_interval;
-			break;
-		case HID_GD_JOYSTICK:
-			if (hid_jspoll_interval > 0)
-				interval = hid_jspoll_interval;
-			break;
-		case HID_GD_KEYBOARD:
-			if (hid_kbpoll_interval > 0)
-				interval = hid_kbpoll_interval;
-			break;
-		}
-
+		 if(hid_poll_interval > 0) {
+			 interval = hid_poll_interval;
+		 } else {
+			switch (hid->collection->usage) {
+			case HID_GD_MOUSE:
+				if (hid_mousepoll_interval > 0)
+					interval = hid_mousepoll_interval;
+				break;
+			case HID_GD_JOYSTICK:
+				if (hid_jspoll_interval > 0)
+					interval = hid_jspoll_interval;
+				break;
+			case HID_GD_KEYBOARD:
+				if (hid_kbpoll_interval > 0)
+					interval = hid_kbpoll_interval;
+				break;
+			}
+		 }
+		 
 		ret = -ENOMEM;
 		if (usb_endpoint_dir_in(endpoint)) {
 			if (usbhid->urbin)
