From b087c6a5ff85f0ca5777d8e89ca98ef031f5f62f Mon Sep 17 00:00:00 2001
From: Vladislav Nepogodin <nepogodin.vlad@gmail.com>
Date: Sat Jan 22 00:52:58 2022 +0300
Subject: [PATCH] GCC: Use #pragma to suppress warnings in GCC truck

Signed-off-by: Vladislav Nepogodin <nepogodin.vlad@gmail.com>
---

diff --git a/tools/lib/subcmd/subcmd-util.h b/tools/lib/subcmd/subcmd-util.h
index 794a375da..075478945 100644
--- a/tools/lib/subcmd/subcmd-util.h
+++ b/tools/lib/subcmd/subcmd-util.h
@@ -49,6 +49,10 @@ static NORETURN inline void die(const char *err, ...)
 
 static inline void *xrealloc(void *ptr, size_t size)
 {
+#if defined(__GNUC__) && !defined(__clang__)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wuse-after-free"
+#endif
 	void *ret = realloc(ptr, size);
 	if (!ret && !size)
 		ret = realloc(ptr, 1);
@@ -59,6 +63,9 @@ static inline void *xrealloc(void *ptr, size_t size)
 		if (!ret)
 			die("Out of memory, realloc failed");
 	}
+#if defined(__GNUC__) && !defined(__clang__)
+#pragma GCC diagnostic pop
+#endif
 	return ret;
 }
 
diff --git a/tools/objtool/check.c b/tools/objtool/check.c
index 21735829b..8c296324e 100644
--- a/tools/objtool/check.c
+++ b/tools/objtool/check.c
@@ -2833,7 +2833,14 @@ static inline const char *call_dest_name(struct instruction *insn)
 	rel = insn_reloc(NULL, insn);
 	if (rel && !strcmp(rel->sym->name, "pv_ops")) {
 		idx = (rel->addend / sizeof(void *));
+#if defined(__GNUC__) && !defined(__clang__)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wformat-truncation"
+#endif
 		snprintf(pvname, sizeof(pvname), "pv_ops[%d]", idx);
+#if defined(__GNUC__) && !defined(__clang__)
+#pragma GCC diagnostic pop
+#endif
 		return pvname;
 	}
 
