--- /src/nvidia/src/kernel/platform/cpu.c    Thu Aug  1 16:06:02 2024
+++ /src/nvidia/src/kernel/platform/cpu.c    Mon Aug 19 18:26:55 2024
@@ -1400,12 +1400,14 @@
                 // Zen, Zen+, Zen 2
             case 0x0A0:
                 // Zen 3, Zen 4
+            case 0x0B0:
+                // Zen 5
                 pSys->cpuInfo.type = NV0000_CTRL_SYSTEM_CPU_TYPE_RYZEN;
                 break;
             default:
-                NV_PRINTF(LEVEL_ERROR,
-                          "Unrecognized AMD processor in cpuidInfoAMD\n");
-                pSys->cpuInfo.type = NV0000_CTRL_SYSTEM_CPU_TYPE_K8;
+                NV_PRINTF(LEVEL_NOTICE,
+                          "Unrecognized AMD processor 0x%x in cpuidInfoAMD. Assuming new Ryzen\n", pCpuidInfo->Family);
+                pSys->cpuInfo.type = NV0000_CTRL_SYSTEM_CPU_TYPE_RYZEN;
                 break;
         }
     }
