From c1ee1e0ba639cdd049b25cfecac9eeb7b3ae68e3 Mon Sep 17 00:00:00 2001
From: Peter Jung <admin@ptr1337.dev>
Date: Mon, 28 Aug 2023 14:05:49 +0200
Subject: [PATCH] Add extra version CachyOS

Signed-off-by: Peter Jung <admin@ptr1337.dev>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 8a601d85cd3f..50ddf3d94445 100644
--- a/Makefile
+++ b/Makefile
@@ -2,7 +2,7 @@
 VERSION = 6
 PATCHLEVEL = 5
 SUBLEVEL = 9
-EXTRAVERSION =
+EXTRAVERSION = -cachyos
 NAME = Hurr durr I'ma ninja sloth
 
 # *DOCUMENTATION*
-- 
2.42.0

