From cc657f4ce50eb33d5e66c364994d35c5395ab441 Mon Sep 17 00:00:00 2001
From: Peter Jung <admin@ptr1337.dev>
Date: Thu, 18 May 2023 19:41:40 +0200
Subject: [PATCH] Add extra version CachyOS

Signed-off-by: Peter Jung <admin@ptr1337.dev>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index f4e38dc60903..881f357f9be7 100644
--- a/Makefile
+++ b/Makefile
@@ -2,7 +2,7 @@
 VERSION = 6
 PATCHLEVEL = 3
 SUBLEVEL = 9
-EXTRAVERSION =
+EXTRAVERSION = -cachyos
 NAME = Hurr durr I'ma ninja sloth
 
 # *DOCUMENTATION*
-- 
2.41.0.rc0

