From c892f6f9e685ca6f83b6aa361138d66d62dd0936 Mon Sep 17 00:00:00 2001
From: Peter Jung <admin@ptr1337.dev>
Date: Sat, 12 Apr 2025 20:39:17 +0200
Subject: [PATCH] kbuild: Add workaround for GCC 15 Compilation

Signed-off-by: Peter Jung <admin@ptr1337.dev>
---
 kernel-open/Kbuild | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/kernel-open/Kbuild b/kernel-open/Kbuild
index 180d53be..8eb29b6b 100644
--- a/kernel-open/Kbuild
+++ b/kernel-open/Kbuild
@@ -81,6 +81,9 @@ ccflags-y += -Wall $(DEFINES) $(INCLUDES) -Wno-cast-qual -Wno-format-extra-args
 ccflags-y += -D__KERNEL__ -DMODULE -DNVRM
 ccflags-y += -DNV_VERSION_STRING=\"570.133.07\"
 
+# GCC 15 workaround
+ccflags-y += -Wno-error=return-type -Wno-implicit-function-declaration -Wno-int-conversion -Wno-incompatible-pointer-types
+
 ifneq ($(SYSSRCHOST1X),)
  ccflags-y += -I$(SYSSRCHOST1X)
 endif
-- 
2.48.0.rc1

