From b6cb15c2e478aecde5a956c5b99b5b50234426d2 Mon Sep 17 00:00:00 2001
From: Peter Jung <admin@ptr1337.dev>
Date: Sat, 1 Jul 2023 15:13:08 +0200
Subject: [PATCH] Add extra version CachyOS

Signed-off-by: Peter Jung <admin@ptr1337.dev>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 5547e02f6..c8e8530de 100644
--- a/Makefile
+++ b/Makefile
@@ -2,7 +2,7 @@
 VERSION = 6
 PATCHLEVEL = 4
 SUBLEVEL = 12
-EXTRAVERSION =
+EXTRAVERSION = -cachyos
 NAME = Hurr durr I'ma ninja sloth
 
 # *DOCUMENTATION*
-- 
2.42.0.rc0.25.ga82fb66fed

