From 335aae5347d9e9d835c240ee0074226eddeeb346 Mon Sep 17 00:00:00 2001
From: Peter Jung <admin@ptr1337.dev>
Date: Fri, 12 Jul 2024 17:31:24 +0200
Subject: [PATCH] BOLT-vmlinux

Signed-off-by: Peter Jung <admin@ptr1337.dev>
---
 scripts/link-vmlinux.sh | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/scripts/link-vmlinux.sh b/scripts/link-vmlinux.sh
index 518c70b8db50..1131a9607a46 100755
--- a/scripts/link-vmlinux.sh
+++ b/scripts/link-vmlinux.sh
@@ -300,5 +300,13 @@ if is_enabled CONFIG_KALLSYMS; then
 	fi
 fi
 
+# Apply BOLT
+BOLT=llvm-bolt
+BOLT_PROFILE=perf.fdata
+BOLT_OPTS="--dyno-stats --eliminate-unreachable=0 --reorder-blocks=ext-tsp --simplify-conditional-tail-calls=0 --skip-funcs=__entry_text_start,irq_entries_start --split-functions"
+mv vmlinux vmlinux.pre-bolt
+echo BOLTing vmlinux
+${BOLT} vmlinux.pre-bolt -o vmlinux --data ${BOLT_PROFILE} ${BOLT_OPTS}
+
 # For fixdep
 echo "vmlinux: $0" > .vmlinux.d
-- 
2.46.0.rc0

