From aeb7b7c10a7d99238ea5130da157580557019808 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Mon, 3 Jun 2024 09:52:03 +0200
Subject: [PATCH] Add extra version CachyOS

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 7f921ae54..1ca9208f9 100644
--- a/Makefile
+++ b/Makefile
@@ -2,7 +2,7 @@
 VERSION = 6
 PATCHLEVEL = 10
 SUBLEVEL = 0
-EXTRAVERSION = -rc2
+EXTRAVERSION = -rc2-cachyos
 NAME = Baby Opossum Posse
 
 # *DOCUMENTATION*
-- 
2.45.1.145.g83f1add914

